opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -11,3 +11,513 @@
+                          ) -> 'GaugeLong':
+        """Creates a gauge type metric that contains long values.
+
+        Args:
+            name: The name of the gauge.
+            options: An optional :class:`.MetricOptions` used to populate
+            details about the gauge.
+
+        Returns:
+            A new :class:`.GaugeLong`
+        """
+
+    def create_measure(self,
+                       name: str,
+                       options: typing.Optional['MeasureOptions'] = None
+                       ) -> 'Measure':
+        """Creates a Measure used to record raw :class:`.Measurement`s.
+
+        Args:
+            name: the name of the measure
+            options: An optional :class:`.MeasureOptions` used to populate
+            details about the measure
+
+        Returns:
+            A :class:`.Measure`
+        """
+
+    def record(self,
+               measurements: typing.List['Measurement'],
+               options: typing.Optional['RecordOptions'] = None) -> None:
+        """A function use to record a set of :class:`.Measurement`s.
+
+        The API is built with the idea that measurement aggregation will occur
+        asynchronously. Typical library records multiple measurements at once,
+        so this function accepts a collection of measurements so the library
+        can batch all of them that need to be recorded.
+
+        Args:
+            measurements: The collection of measurements to record. options:
+            An optional :class:`.RecordOptions` used to populate details during
+            recording.
+
+        Returns: None
+        """
+
+
+_METER: typing.Optional[Meter] = None
+_METER_FACTORY: typing.Optional[
+    typing.Callable[[typing.Type[Meter]], typing.Optional[Meter]]] = None
+
+
+def meter() -> Meter:
+    """Gets the current global :class:`.Meter` object.
+
+    If there isn't one set yet, a default will be loaded.
+    """
+    global _METER, _METER_FACTORY  # pylint:disable=global-statement
+
+    if _METER is None:
+        # pylint:disable=protected-access
+        _METER = loader._load_impl(Meter, _METER_FACTORY)
+        del _METER_FACTORY
+
+    return _METER
+
+
+def set_preferred_meter_implementation(
+        factory: typing.Callable[
+            [typing.Type[Meter]], typing.Optional[Meter]]
+        ) -> None:
+    """Set the factory to be used to create the :class:`.Meter`.
+
+    See :mod:`opentelemetry.loader` for details.
+
+    This function may not be called after a meter is already loaded.
+
+    Args:
+        factory: Callback that should create a new :class:`.Meter` instance.
+    """
+    global _METER_FACTORY  # pylint:disable=global-statement
+
+    if _METER:
+        raise RuntimeError("Meter already loaded.")
+
+    _METER_FACTORY = factory
+
+
+class LabelKey:
+
+    def __init__(self,
+                 key: str,
+                 description: str) -> None:
+        self.key = key
+        self.description = description
+
+
+class LabelValue:
```

Comment:
As I understand it we may pick up entry key/values from the context, but they're label key/values on the metric. Entries get sent between services as headers, but labels are shipped off to backends when we export metrics. In particular, they might not be restricted to 256 chars, or the exporters might be responsible for truncating/dropping long labels instead of the core library.

Label keys and values are part of the export API, they're only defined in the proto now AFAIK: https://github.com/open-telemetry/opentelemetry-proto/blob/d997c83377661c35beb9a8b7535ca00c55f403c4/src/opentelemetry/proto/metrics/v1/metrics.proto#L105.

I'd take the metrics spec as it is with a big helping of salt.